tsl/src/compression/api.c

```diff
@@ -892,3 +976,649 @@ tsl_recompress_chunk_wrapper(Chunk *uncompressed_chunk)
+															colinfo_array,
+															htcols_listlen,
+															&n_keys,
+															&keys);
+
+	/* XXX: Ideally, we want to take RowExclusiveLock to allow more concurrent operations than
+	 * simply SELECTs */
+	Relation compressed_chunk_rel = table_open(compressed_chunk->table_id, ExclusiveLock);
+	Relation uncompressed_chunk_rel = table_open(uncompressed_chunk->table_id, ExclusiveLock);
+
+	/****** compression statistics ******/
+	RelationSize after_size;
+
+	Tuplesortstate *segment_tuplesortstate;
+
+	/*************** tuplesort state *************************/
+	TupleDesc compressed_rel_tupdesc = RelationGetDescr(compressed_chunk_rel);
+	TupleDesc uncompressed_rel_tupdesc = RelationGetDescr(uncompressed_chunk_rel);
+
+	AttrNumber *sort_keys = palloc(sizeof(*sort_keys) * n_keys);
+	Oid *sort_operators = palloc(sizeof(*sort_operators) * n_keys);
+	Oid *sort_collations = palloc(sizeof(*sort_collations) * n_keys);
+	bool *nulls_first = palloc(sizeof(*nulls_first) * n_keys);
+
+	for (int n = 0; n < n_keys; n++)
+		compress_chunk_populate_sort_info_for_column(RelationGetRelid(uncompressed_chunk_rel),
+													 keys[n],
+													 &sort_keys[n],
+													 &sort_operators[n],
+													 &sort_collations[n],
+													 &nulls_first[n]);
+
+	segment_tuplesortstate = tuplesort_begin_heap(uncompressed_rel_tupdesc,
+												  n_keys,
+												  sort_keys,
+												  sort_operators,
+												  sort_collations,
+												  nulls_first,
+												  maintenance_work_mem,
+												  NULL,
+												  false);
+
+	/******************** row decompressor **************/
+
+	RowDecompressor decompressor = build_decompressor(compressed_chunk_rel, uncompressed_chunk_rel);
+	/* do not need the indexes on the uncompressed chunk as we do not write to it anymore */
+	ts_catalog_close_indexes(decompressor.indexstate);
+	/********** row compressor *******************/
+	RowCompressor row_compressor;
+	row_compressor_init(&row_compressor,
+						uncompressed_rel_tupdesc,
+						compressed_chunk_rel,
+						htcols_listlen,
+						colinfo_array,
+						in_column_offsets,
+						compressed_rel_tupdesc->natts,
+						true /*need_bistate*/,
+						true /*recompressing*/);
+
+	HeapTuple compressed_tuple;
+
+	IndexScanDesc index_scan;
+	SegmentInfo *segment_info = NULL;
+	bool changed_segment = false;
+	/************ current segment **************/
+	CompressedSegmentInfo **current_segment =
+		palloc(sizeof(CompressedSegmentInfo *) * nsegmentby_cols);
+
+	for (int i = 0; i < nsegmentby_cols; i++)
+	{
+		current_segment[i] = palloc(sizeof(CompressedSegmentInfo));
+	}
+	bool current_segment_init = false;
+
+	/************** snapshot ****************************/
+	Snapshot snapshot = RegisterSnapshot(GetTransactionSnapshot());
+
+	/* Index scan */
+	elog(DEBUG1, "Index oid for recompress is %u", row_compressor.index_oid);
+	Relation index_rel = index_open(row_compressor.index_oid, AccessExclusiveLock);
```

Comment:
Why do we need AccessExclusiveLock here?

Reply:
basically I did this to avoid the deadlock between two concurrent recompress operations https://github.com/timescale/timescaledb/pull/5321#discussion_r1136639296 but that deadlock might be acceptable?